### ISSUE TITLE
Add DefaultConverterProvider support for conversions between TIME[TAMP]TZ strings and Local[Date]Time

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12703]
We currently cannot directly convert from `String` to `Timestamp`/`Time` or `LocalDateTime`/`LocalTime`, if the string version of the timestamp has a time zone:

```java
ConverterProvider cp = new DefaultConverterProvider();

// This prints null
Converter<String, LocalDateTime> c = cp.provide(String.class, LocalDateTime.class);
System.out.println(c.from("2000-01-01 10:00:00+02"));
```

However, if we convert first to `OffsetDateTime`/`OffsetTime`, then it works:

```java
// This prints 2000-01-01T09:00, when running in CET
Converter<String, OffsetDateTime> c1 = cp.provide(String.class, OffsetDateTime.class);
Converter<OffsetDateTime, LocalDateTime> c2 = cp.provide(OffsetDateTime.class, LocalDateTime.class);
System.out.println(c1.andThen(c2).from("2000-01-01 10:00:00+02"));
```

We've been patching the internal temporal parsing logic time and again, it seems that these problems can only be fixed thoroughly by completely re-implementing parsing on our own, rather than patching things all the time.

This has already been done for performance reasons, because the `java.time` parser is very slow, even for the very popular ISO 8601 formats! See #7986, https://bugs.openjdk.java.net/browse/JDK-8213243. With just a little more effort, we can retrofit this parser for all temporal data types.

**Comment by lukaseder** [comment]
This should fix https://github.com/jOOQ/jOOQ/issues/12699

**Comment by katzyn** [comment]
I'm just wondering how this client-side conversion can take time zone of database session or transaction into account? Even if driver knows it internally, it isn't exposed via JDBC API.

**Comment by katzyn** [comment]
`Timestamp` internally holds UTC-based value, similar to `java.time.Instant`, but with various bugs. Conversion from `TIMESTAMP WITH TIME ZONE` to `java.sql.Timestamp` doesn't depend on time zones at all. But conversion to `LocalDateTime` does.

**Comment by lukaseder** [comment]
Yes that's right. The epoch timestamp remains unchanged, but `Timestamp::toString`, `Timestamp::toLocalDateTime` as well as both `Timestamp::valueOf` use the client JVM's default time zone, so in the context of this task here, is there really any issue?

**Comment by katzyn** [comment]
`ResultSet.get(column, LocalDateTime.class)` with properly written drivers doesn't use `Timestamp` at all, but constructs this object from original value received from database and most likely it will be in time zone of session, transaction, or whatever database has.

**Comment by lukaseder** [comment]
Which RDBMS do that?

**Comment by katzyn** [comment]
For example, in PostgreSQL conversion depends on time zone of current session or transaction:
```Java
Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost/postgres", "postgres", "");
Statement stat = conn.createStatement();
for (int i = 0; i < 10; i++) {
    stat.executeUpdate("SET TIME ZONE '" + i + ":00'");
    // Cast is needed here to avoid exception in PgResultSet.getLocalDateTime()
    ResultSet rs = stat.executeQuery(
            "SELECT CAST(TIMESTAMP WITH TIME ZONE '2001-01-01 00:00:00+00' AS TIMESTAMP)");
    rs.next();
    System.out.println(rs.getObject(1, LocalDateTime.class));
}
```

```
2001-01-01T00:00
2000-12-31T23:00
2000-12-31T22:00
2000-12-31T21:00
2000-12-31T20:00
2000-12-31T19:00
2000-12-31T18:00
2000-12-31T17:00
2000-12-31T16:00
2000-12-31T15:00
```

In H2 it also depends on time zone of the session:
```Java
Connection conn = DriverManager.getConnection("jdbc:h2:mem:");
Statement stat = conn.createStatement();
for (int i = 0; i < 10; i++) {
    stat.executeUpdate("SET TIME ZONE '" + i + ":00'");
    // Cast isn't required
    ResultSet rs = stat.executeQuery("SELECT TIMESTAMP WITH TIME ZONE '2001-01-01 00:00:00+00'");
    rs.next();
    System.out.println(rs.getObject(1, LocalDateTime.class));
}
```
(Results are the same.)

**Comment by lukaseder** [comment]
> // Cast is needed here to avoid exception in PgResultSet.getLocalDateTime()

To avoid the exception, but you changed the semantics! Try this in your PostgreSQL loop:

```java
System.out.println(rs.getObject(1, LocalDateTime.class));
System.out.println(rs.getTimestamp(1));
```

You'll get:

```
2001-01-01T00:00
2001-01-01 00:00:00.0
2000-12-31T23:00
2000-12-31 23:00:00.0
2000-12-31T22:00
2000-12-31 22:00:00.0
2000-12-31T21:00
2000-12-31 21:00:00.0
2000-12-31T20:00
2000-12-31 20:00:00.0
2000-12-31T19:00
2000-12-31 19:00:00.0
2000-12-31T18:00
2000-12-31 18:00:00.0
2000-12-31T17:00
2000-12-31 17:00:00.0
2000-12-31T16:00
2000-12-31 16:00:00.0
2000-12-31T15:00
2000-12-31 15:00:00.0
```

**Comment by lukaseder** [comment]
Note again the duality of the `java.sql.Timestamp` ugly data type, which can be used to model both `TIMESTAMP` (via `Timestamp::toString` and `Timestamp::valueOf`) and `TIMESTAMPTZ` (via `Timestamp::getTime` and `Timestamp::new`).

In your case, your `CAST` is significant and changed the semantics. It's not just to avoid an exception. It seems that pgjdbc decided not to do any such client timezone magic, though I think it could have...

**Comment by lukaseder** [comment]
Yeah, no one doubts it's a mess, especially on systems that for whatever reason choose to use different client and server time zones. Though I personally found plain `getTimestamp()` to be reliable enough in most cases.

Once we have the time zone SPI (#11761), we can use the `Calendar` using overloads. Before that (and hardly anyone has ever asked for it), `getTimestamp()` is good enough.

Which brings me back to this `DefaultConverterProvider` task, which will have access to the time zone SPI, but can already now assume the default SPI implementation will be using the JVM's default time zone. What will be the problem then, if an API that is purely client based, will convert between these data types?
